ext/json/json_encoder.c

```diff
@@ -104,7 +104,11 @@ static inline void php_json_encode_double(smart_str *buf, double d, int options)
 {
 	size_t len;
 	char num[PHP_JSON_DOUBLE_MAX_LENGTH];
-	php_gcvt(d, (int)EG(precision), '.', 'e', &num[0]);
+	if (JSON_G(precision) == -1) {
+		php_0cvt(d, 17, '.', 'e', num);
+	} else {
+		php_gcvt(d, JSON_G(precision), '.', 'e', num);
+	}
```

Comment:
Good point, looks like dtoa doesn't directly deal with stuff like scientific notation. Given that the dtoa 0 mode (or precision -1 mode) is supposed to return an *accurate* result, I don't think it's necessary to have separate ini settings for this -- we should simply automatically choose an appropriate ndigits for these checks if it was -1 originally. IIRC ndigits=17 is enough for an accurace result and I don't think this will produce overlong results, but not sure on that count. This will need some testing...